File: cli/internal/webdav.go
package internal

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"path"
	"path/filepath"
	"strings"
	"time"
)

// ErrRemoteNotFound is returned when the remote file does not exist (HTTP 404).
var ErrRemoteNotFound = errors.New("remote file not found")

// ErrEtagMismatch is returned when a conditional PUT fails because the remote
// resource changed under us (HTTP 412 Precondition Failed). Callers can react
// by re-fetching and merging or by surfacing the conflict to the user.
var ErrEtagMismatch = errors.New("remote etag changed since last fetch")

// IfNoneMatchAny is the value used as the ifMatch argument when the caller
// wants the PUT to succeed only if the resource does not yet exist (i.e.
// translated by the WebDAV layer to an If-None-Match: * header).
const IfNoneMatchAny = "*"

// ErrWebDAVHandledByDesktopClient is returned when the local data file lives
// inside a Nextcloud desktop-client sync folder, so the CLI should not also
// push to WebDAV — doing both would race and create "conflicted copy" files.
var ErrWebDAVHandledByDesktopClient = errors.New("local data file is inside a Nextcloud sync folder; desktop client handles sync")

// LocalPathInNextcloudSyncFolder reports whether dataPath resolves to a
// location inside the user's Nextcloud desktop-client sync folder (~/Nextcloud).
// When true, the desktop client already propagates writes to the server, and
// the CLI should not perform its own WebDAV PUTs against the same file.
func LocalPathInNextcloudSyncFolder(dataPath string) bool {
	if dataPath == "" {
		return false
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return false
	}
	abs, err := filepath.Abs(dataPath)
	if err != nil {
		return false
	}
	rel, err := filepath.Rel(filepath.Join(home, "Nextcloud"), abs)
	if err != nil {
		return false
	}
	return rel != "." && !strings.HasPrefix(rel, "..")
}

// WebDAVSettings contains the configuration for WebDAV sync
type WebDAVSettings struct {
	URL  string
	User string
	Pass string
}

func buildWebDAVURL(cfg NextcloudConfig) string {
	base := strings.TrimRight(cfg.ServerURL, "/")
	path := cfg.RemotePath
	if !strings.HasPrefix(path, "/") {
		path = "/" + path
	}
	return base + path
}

func buildHistoryWebDAVURL(raw string) string {
	parsed, err := url.Parse(raw)
	if err != nil {
		return raw + ".history.json"
	}
	ext := path.Ext(parsed.Path)
	if ext == "" {
		parsed.Path += ".history.json"
	} else {
		base := strings.TrimSuffix(parsed.Path, ext)
		parsed.Path = base + ".history" + ext
	}
	return parsed.String()
}

// LoadWebDAVSettings loads WebDAV configuration from the persisted backend config.
func LoadWebDAVSettings() (WebDAVSettings, error) {
	cfg, _, err := LoadEffectiveAppConfig()
	if err != nil {
		return WebDAVSettings{}, err
	}
	if cfg.Backend != BackendNextcloud || !nextcloudConfigComplete(cfg.Nextcloud) {
		return WebDAVSettings{}, ErrNextcloudNotConfigured
	}
	return WebDAVSettings{
		URL:  buildWebDAVURL(cfg.Nextcloud),
		User: cfg.Nextcloud.LoginName,
		Pass: cfg.Nextcloud.AppPassword,
	}, nil
}

// FetchRemoteData fetches the data from the remote WebDAV server and returns
// the response ETag alongside the parsed body. The etag is opaque and should
// be passed back verbatim to PushRemoteData as ifMatch on a subsequent PUT
// to detect concurrent writers (412 → ErrEtagMismatch).
func FetchRemoteData(settings WebDAVSettings) (Data, string, error) {
	client := &http.Client{Timeout: 15 * time.Second}
	req, err := http.NewRequest(http.MethodGet, settings.URL, nil)
	if err != nil {
		return Data{}, "", err
	}
	req.SetBasicAuth(settings.User, settings.Pass)
	resp, err := client.Do(req)
	if err != nil {
		return Data{}, "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusNotFound {
		return Data{}, "", ErrRemoteNotFound
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return Data{}, "", fmt.Errorf("remote status %d", resp.StatusCode)
	}
	var data Data
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return Data{}, "", err
	}
	return data, resp.Header.Get("ETag"), nil
}

// PushRemoteData pushes data to the remote WebDAV server. ifMatch carries the
// ETag from a prior FetchRemoteData call: pass "" to push unconditionally,
// IfNoneMatchAny to require the resource not yet exist (initial create), or
// any other value to require the resource still match that ETag. On a 412
// Precondition Failed response this returns ErrEtagMismatch so callers can
// refetch and merge instead of silently clobbering a concurrent writer.
//
// The caller is expected to have set data.LastModified (normally via
// SaveData). This function does not restamp it: doing so would make the
// bytes pushed differ from the bytes on disk, which would itself cause
// content-level conflicts with any Nextcloud desktop client syncing the
// same file.
func PushRemoteData(settings WebDAVSettings, data Data, ifMatch string) error {
	if data.LastModified == 0 {
		data.LastModified = time.Now().UnixMilli()
	}
	body, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}

	client := &http.Client{Timeout: 15 * time.Second}
	req, err := http.NewRequest(http.MethodPut, settings.URL, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	setIfMatch(req, ifMatch)
	req.SetBasicAuth(settings.User, settings.Pass)

	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusPreconditionFailed {
		return ErrEtagMismatch
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("push failed with status %d", resp.StatusCode)
	}
	return nil
}

// setIfMatch translates an ifMatch argument into the corresponding precondition
// header. The IfNoneMatchAny sentinel maps to If-None-Match: * (only allow
// create); any other non-empty value goes onto If-Match verbatim.
func setIfMatch(req *http.Request, ifMatch string) {
	switch ifMatch {
	case "":
		// no precondition
	case IfNoneMatchAny:
		req.Header.Set("If-None-Match", "*")
	default:
		req.Header.Set("If-Match", ifMatch)
	}
}

// FetchRemoteHistory returns the parsed history payload along with the ETag
// for the resource (empty if the server did not provide one). Pass the etag
// back to PushRemoteHistory as ifMatch on a subsequent PUT to detect
// concurrent writers.
func FetchRemoteHistory(settings WebDAVSettings) (History, string, error) {
	client := &http.Client{Timeout: 15 * time.Second}
	req, err := http.NewRequest(http.MethodGet, buildHistoryWebDAVURL(settings.URL), nil)
	if err != nil {
		return History{}, "", err
	}
	req.SetBasicAuth(settings.User, settings.Pass)
	resp, err := client.Do(req)
	if err != nil {
		return History{}, "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusNotFound {
		return History{}, "", ErrRemoteNotFound
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return History{}, "", fmt.Errorf("remote history status %d", resp.StatusCode)
	}
	var history History
	if err := json.NewDecoder(resp.Body).Decode(&history); err != nil {
		return History{}, "", err
	}
	if history.Version == 0 {
		history.Version = historyVersion
	}
	sortHistory(&history)
	return history, resp.Header.Get("ETag"), nil
}

// PushRemoteHistory pushes the history payload. The ifMatch argument follows
// the same conventions as PushRemoteData: "" for unconditional, IfNoneMatchAny
// for create-only, otherwise an opaque ETag from a prior fetch.
// Returns ErrEtagMismatch on 412.
func PushRemoteHistory(settings WebDAVSettings, history History, ifMatch string) error {
	history.Version = historyVersion
	if history.UpdatedAt == 0 {
		history.UpdatedAt = time.Now().UnixMilli()
	}
	sortHistory(&history)

	body, err := json.MarshalIndent(history, "", "  ")
	if err != nil {
		return err
	}

	client := &http.Client{Timeout: 15 * time.Second}
	req, err := http.NewRequest(http.MethodPut, buildHistoryWebDAVURL(settings.URL), bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	setIfMatch(req, ifMatch)
	req.SetBasicAuth(settings.User, settings.Pass)

	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusPreconditionFailed {
		return ErrEtagMismatch
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("history push failed with status %d", resp.StatusCode)
	}
	return nil
}

// PushRemoteState pushes both data and history unconditionally — used by the
// dedicated "push" command (`daily-tasks push`, the TUI push action). It does
// not check the remote etag; it deliberately overwrites whatever is on the
// server. Sync flows go through SyncWithRemote / SyncStateWithRemote, which
// fetch first and use If-Match to avoid clobbering concurrent writers.
func PushRemoteState(settings WebDAVSettings, data Data, history History) error {
	if err := PushRemoteData(settings, data, ""); err != nil {
		return err
	}
	return PushRemoteHistory(settings, HistoryWithCurrentSnapshot(history, data, time.Now().UnixMilli()), "")
}

// SyncRemoteHistory fetches the remote history, merges it with the supplied
// local history (and current snapshot), and pushes the merged result back if
// it differs from what the server has. Pushes are conditional on the ETag
// captured at fetch time; on ErrEtagMismatch the function refetches and
// retries once, picking up whatever a concurrent writer just deposited.
func SyncRemoteHistory(settings WebDAVSettings, local History, current Data) (History, error) {
	merged, err := mergeAndPushHistory(settings, local, current)
	if err == nil || !errors.Is(err, ErrEtagMismatch) {
		return merged, err
	}
	// Etag changed mid-sync: another writer pushed between our fetch and
	// our PUT. Retry once with the latest remote state folded in.
	return mergeAndPushHistory(settings, local, current)
}

func mergeAndPushHistory(settings WebDAVSettings, local History, current Data) (History, error) {
	remote, etag, err := FetchRemoteHistory(settings)
	notFound := errors.Is(err, ErrRemoteNotFound)
	if err != nil && !notFound {
		return History{}, err
	}
	if notFound {
		remote = History{Version: historyVersion}
	}

	merged := HistoryWithCurrentSnapshot(MergeHistories(local, remote), current, time.Now().UnixMilli())

	if notFound {
		if len(merged.Days) > 0 || len(merged.Events) > 0 {
			if pushErr := PushRemoteHistory(settings, merged, IfNoneMatchAny); pushErr != nil {
				return History{}, pushErr
			}
		}
		return merged, nil
	}

	if !HistoryContentEqual(merged, remote) {
		if pushErr := PushRemoteHistory(settings, merged, etag); pushErr != nil {
			return History{}, pushErr
		}
	}
	return merged, nil
}

// SyncResult represents the result of a sync operation
type SyncResult struct {
	Data     Data
	Action   string // "pulled", "pushed", "in_sync", "error"
	Message  string
	Conflict bool
}

type SyncStateResult struct {
	Data    Data
	History History
	Action  string
	Message string
}

// SyncWithRemote performs a bi-directional sync with conflict detection.
// Pushes are conditional on the ETag captured at fetch time, so a concurrent
// writer can be detected (412) instead of silently clobbered. On
// ErrEtagMismatch the function refetches once and re-evaluates: typically the
// remote is now newer, so the result becomes a pull rather than another push.
func SyncWithRemote(settings WebDAVSettings, local Data) SyncResult {
	result, retry := syncOnce(settings, local)
	if !retry {
		return result
	}
	result, _ = syncOnce(settings, local)
	return result
}

// syncOnce runs a single sync attempt. The bool return is true when the
// caller should retry (because a concurrent writer changed the etag between
// our fetch and our PUT).
func syncOnce(settings WebDAVSettings, local Data) (SyncResult, bool) {
	remote, etag, err := FetchRemoteData(settings)
	if err != nil {
		// If remote doesn't exist, push local with If-None-Match: *
		// so we don't clobber a file another client just created.
		if errors.Is(err, ErrRemoteNotFound) {
			pushErr := PushRemoteData(settings, local, IfNoneMatchAny)
			if errors.Is(pushErr, ErrEtagMismatch) {
				// Lost the create race — retry path will pull or merge.
				return SyncResult{Data: local, Action: "error", Message: "Remote was created by another writer; retrying"}, true
			}
			if pushErr != nil {
				return SyncResult{Data: local, Action: "error", Message: fmt.Sprintf("Push failed: %s", pushErr)}, false
			}
			return SyncResult{Data: local, Action: "pushed", Message: "Created remote file"}, false
		}
		return SyncResult{Data: local, Action: "error", Message: err.Error()}, false
	}

	remote = NormalizeData(remote)
	local = NormalizeData(local)

	// Never overwrite remote tasks with an empty local state (e.g. fresh install)
	if len(local.Tasks) == 0 && len(remote.Tasks) > 0 {
		return SyncResult{Data: remote, Action: "pulled", Message: "Pulled remote data"}, false
	}

	// Check for conflicts using LastModified
	if remote.LastModified > local.LastModified {
		// Remote is newer, pull it
		return SyncResult{Data: remote, Action: "pulled", Message: "Pulled newer remote data"}, false
	} else if local.LastModified > remote.LastModified {
		// Local is newer, push it conditionally on the etag we just saw.
		pushErr := PushRemoteData(settings, local, etag)
		if errors.Is(pushErr, ErrEtagMismatch) {
			return SyncResult{Data: local, Action: "error", Message: "Remote changed during sync; retrying"}, true
		}
		if pushErr != nil {
			return SyncResult{Data: local, Action: "error", Message: fmt.Sprintf("Push failed: %s", pushErr)}, false
		}
		return SyncResult{Data: local, Action: "pushed", Message: "Pushed local changes"}, false
	}

	// Same timestamp, no action needed
	return SyncResult{Data: local, Action: "in_sync", Message: "Already in sync"}, false
}

func SyncStateWithRemote(settings WebDAVSettings, local Data, history History) SyncStateResult {
	result := SyncWithRemote(settings, local)
	if result.Action == "error" {
		return SyncStateResult{
			Data:    result.Data,
			History: history,
			Action:  result.Action,
			Message: result.Message,
		}
	}

	mergedHistory, err := SyncRemoteHistory(settings, history, NormalizeData(result.Data))
	if err != nil {
		return SyncStateResult{
			Data:    result.Data,
			History: history,
			Action:  "error",
			Message: err.Error(),
		}
	}

	return SyncStateResult{
		Data:    NormalizeData(result.Data),
		History: mergedHistory,
		Action:  result.Action,
		Message: result.Message,
	}
}

// HasWebDAVConfig returns true if WebDAV is configured
func HasWebDAVConfig() bool {
	_, err := LoadWebDAVSettings()
	return err == nil && !errors.Is(err, ErrNextcloudNotConfigured)
}

File: mobile/src/services/webdav.ts
import { encode as base64Encode } from 'base-64';
import type { Data, History, Settings } from '../types';
import { normalizeData } from './data';
import { emptyHistory, ensureHistorySnapshot, historyContentEqual, mergeHistories, normalizeHistory } from './history';
import { defaultRemotePath } from '../config/env';

export type LoginFlowSession = {
  serverUrl: string;
  loginUrl: string;
  pollEndpoint: string;
  pollToken: string;
};

export const defaultSettings: Settings = {
  baseUrl: '',
  username: '',
  password: '',
  remotePath: defaultRemotePath,
};

function normalizeServerUrl(serverUrl: string): string {
  return serverUrl.trim().replace(/\/+$/, '');
}

export function isSettingsComplete(settings: Settings): boolean {
  return !!(settings.baseUrl && settings.username && settings.password && settings.remotePath);
}

function basicAuthHeader(settings: Settings): string {
  const token = `${settings.username}:${settings.password}`;
  const encoded = globalThis.btoa ? globalThis.btoa(token) : base64Encode(token);
  return `Basic ${encoded}`;
}

function buildWebdavUrl(settings: Settings): string {
  const base = settings.baseUrl.replace(/\/+$/, '');
  const path = settings.remotePath.startsWith('/') ? settings.remotePath : `/${settings.remotePath}`;
  return `${base}${path}`;
}

function buildHistoryRemotePath(remotePath: string): string {
  const slash = remotePath.lastIndexOf('/');
  const dir = slash >= 0 ? remotePath.slice(0, slash + 1) : '';
  const name = slash >= 0 ? remotePath.slice(slash + 1) : remotePath;
  const dot = name.lastIndexOf('.');
  if (dot <= 0) {
    return `${dir}${name}.history.json`;
  }
  return `${dir}${name.slice(0, dot)}.history${name.slice(dot)}`;
}

function buildHistoryWebdavUrl(settings: Settings): string {
  const base = settings.baseUrl.replace(/\/+$/, '');
  const historyPath = buildHistoryRemotePath(settings.remotePath);
  const normalizedPath = historyPath.startsWith('/') ? historyPath : `/${historyPath}`;
  return `${base}${normalizedPath}`;
}

function buildRemotePath(loginName: string): string {
  return `/remote.php/dav/files/${encodeURIComponent(loginName)}/.daily-tasks.json`;
}

export async function startNextcloudLogin(serverUrl: string): Promise<LoginFlowSession> {
  const normalizedServerUrl = normalizeServerUrl(serverUrl);
  if (!normalizedServerUrl) {
    throw new Error('Server URL is required');
  }

  const res = await fetch(`${normalizedServerUrl}/index.php/login/v2`, {
    method: 'POST',
  });
  if (!res.ok) {
    throw new Error(`Login flow start failed: ${res.status}`);
  }

  const payload = await res.json() as {
    poll?: { token?: string; endpoint?: string };
    login?: string;
  };

  if (!payload.login || !payload.poll?.token || !payload.poll.endpoint) {
    throw new Error('Login flow response was incomplete');
  }

  return {
    serverUrl: normalizedServerUrl,
    loginUrl: payload.login,
    pollEndpoint: payload.poll.endpoint,
    pollToken: payload.poll.token,
  };
}

export async function pollNextcloudLogin(session: LoginFlowSession): Promise<Settings | null> {
  const params = new URLSearchParams();
  params.set('token', session.pollToken);

  const res = await fetch(session.pollEndpoint, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: params.toString(),
  });

  if (res.status === 404) {
    return null;
  }
  if (!res.ok) {
    throw new Error(`Login flow poll failed: ${res.status}`);
  }

  const payload = await res.json() as {
    server?: string;
    loginName?: string;
    appPassword?: string;
  };

  if (!payload.server || !payload.loginName || !payload.appPassword) {
    throw new Error('Login flow poll response was incomplete');
  }

  return {
    baseUrl: normalizeServerUrl(payload.server),
    username: payload.loginName.trim(),
    password: payload.appPassword.trim(),
    remotePath: buildRemotePath(payload.loginName.trim()),
  };
}

// Sentinel passed to pushRemoteData / pushRemoteHistory as ifMatch when the
// caller wants the PUT to succeed only if the resource does not yet exist
// (translated to an If-None-Match: * request header).
export const IF_NONE_MATCH_ANY = '*';

// Thrown by pushRemoteData / pushRemoteHistory when the server returns
// 412 Precondition Failed — i.e. the resource changed since the etag we
// passed in If-Match was issued. Callers can catch this to refetch and merge.
export class EtagMismatchError extends Error {
  constructor(message = 'Remote etag changed since last fetch') {
    super(message);
    this.name = 'EtagMismatchError';
  }
}

export type FetchedData = { data: Data | null; etag: string };
export type FetchedHistory = { history: History | null; etag: string };

function applyIfMatch(headers: Record<string, string>, ifMatch?: string): void {
  if (!ifMatch) {
    return;
  }
  if (ifMatch === IF_NONE_MATCH_ANY) {
    headers['If-None-Match'] = '*';
  } else {
    headers['If-Match'] = ifMatch;
  }
}

export async function fetchRemoteData(settings: Settings): Promise<FetchedData> {
  const url = buildWebdavUrl(settings);
  const res = await fetch(url, {
    headers: {
      Authorization: basicAuthHeader(settings),
      Accept: 'application/json',
    },
  });
  if (res.status === 404) {
    return { data: null, etag: '' };
  }
  if (!res.ok) {
    throw new Error(`Fetch failed: ${res.status}`);
  }
  const data = (await res.json()) as Data;
  return { data, etag: res.headers.get('ETag') ?? '' };
}

export async function fetchRemoteHistory(settings: Settings): Promise<FetchedHistory> {
  const url = buildHistoryWebdavUrl(settings);
  const res = await fetch(url, {
    headers: {
      Authorization: basicAuthHeader(settings),
      Accept: 'application/json',
    },
  });
  if (res.status === 404) {
    return { history: null, etag: '' };
  }
  if (!res.ok) {
    throw new Error(`History fetch failed: ${res.status}`);
  }
  return { history: normalizeHistory(await res.json()), etag: res.headers.get('ETag') ?? '' };
}

// pushRemoteData PUTs the data to the remote WebDAV server. Pass `ifMatch`
// to make the request conditional: an opaque ETag to require the resource
// still match it (If-Match), or IF_NONE_MATCH_ANY to require the resource
// not yet exist (If-None-Match: *). Throws EtagMismatchError on 412.
export async function pushRemoteData(settings: Settings, data: Data, ifMatch?: string): Promise<void> {
  const url = buildWebdavUrl(settings);
  const dataWithTimestamp = {
    ...data,
    last_modified: data.last_modified || Date.now(),
  };
  const headers: Record<string, string> = {
    Authorization: basicAuthHeader(settings),
    'Content-Type': 'application/json',
  };
  applyIfMatch(headers, ifMatch);
  const res = await fetch(url, {
    method: 'PUT',
    headers,
    body: JSON.stringify(dataWithTimestamp, null, 2),
  });
  if (res.status === 412) {
    throw new EtagMismatchError();
  }
  if (!res.ok) {
    throw new Error(`Save failed: ${res.status}`);
  }
}

export async function pushRemoteHistory(settings: Settings, history: History, ifMatch?: string): Promise<void> {
  const url = buildHistoryWebdavUrl(settings);
  const normalized = normalizeHistory(history);
  const payload = {
    ...normalized,
    updated_at: normalized.updated_at || Date.now(),
  };
  const headers: Record<string, string> = {
    Authorization: basicAuthHeader(settings),
    'Content-Type': 'application/json',
  };
  applyIfMatch(headers, ifMatch);
  const res = await fetch(url, {
    method: 'PUT',
    headers,
    body: JSON.stringify(payload, null, 2),
  });
  if (res.status === 412) {
    throw new EtagMismatchError();
  }
  if (!res.ok) {
    throw new Error(`History save failed: ${res.status}`);
  }
}

// pushRemoteState writes both data and history unconditionally — used by the
// dedicated "Save to Nextcloud" flow that is meant to overwrite. Sync flows
// go through syncWithRemote / syncWithRemoteState instead, which fetch first
// and use If-Match to avoid clobbering concurrent writers.
export async function pushRemoteState(settings: Settings, data: Data, history: History): Promise<void> {
  await pushRemoteData(settings, data);
  await pushRemoteHistory(settings, ensureHistorySnapshot(history, data));
}

export async function syncRemoteHistory(settings: Settings, localHistory: History, currentData: Data): Promise<History> {
  try {
    return await mergeAndPushHistory(settings, localHistory, currentData);
  } catch (err) {
    if (err instanceof EtagMismatchError) {
      // Remote etag moved between our fetch and PUT — retry once with the
      // latest server state folded in.
      return mergeAndPushHistory(settings, localHistory, currentData);
    }
    throw err;
  }
}

async function mergeAndPushHistory(
  settings: Settings,
  localHistory: History,
  currentData: Data,
): Promise<History> {
  const local = normalizeHistory(localHistory);
  const { history: remote, etag } = await fetchRemoteHistory(settings);
  const merged = ensureHistorySnapshot(
    mergeHistories(local, remote ?? emptyHistory()),
    currentData,
  );

  if (!remote) {
    if (merged.days.length > 0 || merged.events.length > 0) {
      await pushRemoteHistory(settings, merged, IF_NONE_MATCH_ANY);
    }
    return merged;
  }

  if (!historyContentEqual(merged, remote)) {
    await pushRemoteHistory(settings, merged, etag);
  }

  return merged;
}

export type SyncResult = {
  data: Data;
  action: 'pulled' | 'pushed' | 'error' | 'in_sync';
  message: string;
};

export type SyncStateResult = SyncResult & {
  history: History;
};

async function syncOnce(settings: Settings, localData: Data): Promise<SyncResult> {
  const { data: remoteRaw, etag } = await fetchRemoteData(settings);

  if (!remoteRaw) {
    await pushRemoteData(settings, localData, IF_NONE_MATCH_ANY);
    return { data: localData, action: 'pushed', message: 'Created remote file' };
  }

  const remote = normalizeData(remoteRaw);
  const local = normalizeData(localData);

  // Never overwrite remote tasks with an empty local state (e.g. fresh install or daily reset with no tasks)
  if (local.tasks.length === 0 && remote.tasks.length > 0) {
    return { data: remote, action: 'pulled', message: 'Pulled remote data' };
  }

  const localTimestamp = local.last_modified || 0;
  const remoteTimestamp = remote.last_modified || 0;

  if (remoteTimestamp > localTimestamp) {
    return { data: remote, action: 'pulled', message: 'Pulled newer remote data' };
  } else if (localTimestamp > remoteTimestamp) {
    await pushRemoteData(settings, local, etag);
    return { data: local, action: 'pushed', message: 'Pushed local changes' };
  }

  return { data: local, action: 'in_sync', message: 'Already in sync' };
}

export async function syncWithRemote(settings: Settings, localData: Data): Promise<SyncResult> {
  try {
    return await syncOnce(settings, localData);
  } catch (err) {
    if (err instanceof EtagMismatchError) {
      try {
        return await syncOnce(settings, localData);
      } catch (retryErr) {
        return {
          data: localData,
          action: 'error',
          message: `Sync error: ${(retryErr as Error).message}`,
        };
      }
    }
    return {
      data: localData,
      action: 'error',
      message: `Sync error: ${(err as Error).message}`,
    };
  }
}

export async function syncWithRemoteState(
  settings: Settings,
  localData: Data,
  localHistory: History,
): Promise<SyncStateResult> {
  const result = await syncWithRemote(settings, localData);
  if (result.action === 'error') {
    return { ...result, history: normalizeHistory(localHistory) };
  }

  const normalizedData = normalizeData(result.data);
  const history = await syncRemoteHistory(settings, localHistory, normalizedData);
  return {
    ...result,
    data: normalizedData,
    history,
  };
}
